### PR TITLE
Adding org foreign key to users table

### PR DESCRIPTION
**Pull request by LizPrescott**
This will allow us to continue work on #63 

### Description

  - In order to have the organization's inventory show up as soon as a user logs in, we will need to associate them to an organization
  - TODO: We need to decide what to do with users when they first sign up. Will the homepage be a list of organizations to join if they don't already have one?

### Type of change

* New feature (non-breaking change which adds functionality)

### How Has This Been Tested?

Migrated, rolled back, and then migrated again to make doubly sure there was nothing strange about the migration. I don't think there's any reason to unit test this change.

### CODE REVIEW

**Comment by abronkema** [comment]
@valeriecodes - would you be able to give this a quick review?

File: app/models/user.rb [hunk 1]
```diff
@@ -9,6 +9,7 @@ class User < ApplicationRecord
 
   has_many :signed_agreements
   has_many :carts
+<<<<<<< HEAD
```
Comment (pollygee) [inline app/models/user.rb]: Look like you've still got merge conflict stuff lingering in there?

**Comment by LizPrescott** [comment]
Whops! I will fix that....

File: app/models/user.rb (hunk 1 above)
Comment (LizPrescott) [reply app/models/user.rb]: Fixed!

File: config/routes.rb [hunk 2]
```diff
@@ -13,5 +13,5 @@
 
   resources :categories
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
-  root 'home#index'
```
Comment (valeriecodes) [inline config/routes.rb]: Is this related to the DB migration?

**Review by valeriecodes — COMMENTED** [review]
Looks good! I just had one clarifying question.

File: config/routes.rb (hunk 2 above)
Comment (LizPrescott) [reply config/routes.rb]: Thanks for catching that. It should go in my next pr instead. Should be fixed now.

**Comment by pollygee** [comment]
@valeriecodes is this good to get merged?

File: db/migrate/20191003014319_add_organization_key_to_user_table.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddOrganizationKeyToUserTable < ActiveRecord::Migration[5.2]
```
Comment (dixpac) [inline db/migrate/20191003014319_add_organization_key_to_user_table.rb:1]: I'm not 100% familiar with the project but I would go with the default Rails way of adding foreign_key here:
`add_reference :users, :organization, foreign_key: true`

This is gonna add `organization_id` column and create `foreign_key` constraint out of the box. Foreign key is going to be helpful to the database for the performance and to maintain the "correct" state(less important here since you have `nulls`).

Also, you can get this with the generator like this:
`bin/rails g migration add_organization_key_to_user_table organization:references`

**Comment by LizPrescott** [comment]
Thanks. I’m actually likely closing this since this doesn’t seem like the direction we want to go, but my understanding is that a foreign key constraint would not be appropriate here since users would need to be able to sign up without a parent organization already in place.

**Comment by dixpac** [comment]
@LizPrescott yeah in this case referential integrity is less important since there could be `nulls`, but it will still work since you've specified `optional: true` inside the `User` model, and you would create `reference` with `null: true`

`add_reference :users, :organization, null: true, foreign_key: true`

But more importantly Rails `add_reference` will create you an index on `oganization_id` column, that could be used for faster lookup in queries such as
`User.where(organization: @some_organization)` or `Organization.first.users`

Screenshot 2019-10-09 at 09 50 35


I think that generally it is a good practice to have an index on fk columns.

**Comment by LizPrescott** [comment]
Thanks. Closing this one, but thank you for the indexing reminder.

**Comment by LizPrescott** [comment]
jk jk. I see that in the new description of #63  that this is still relevant. Implementing your suggestion now @dixpac .

File: db/migrate/20191003014319_add_organization_key_to_user_table.rb (hunk 3 above)
Comment (LizPrescott) [reply db/migrate/20191003014319_add_organization_key_to_user_table.rb:1]: > @LizPrescott yeah in this case referential integrity is less important since there could be `nulls`, but it will still work since you've specified `optional: true` inside the `User` model, and you would create `reference` with `null: true`
> 
> `add_reference :users, :organization, null: true, foreign_key: true`
> 
> But more importantly Rails `add_reference` will create you an index on `oganization_id` column, that could be used for faster lookup in queries such as
> `User.where(organization: @some_organization)` or `Organization.first.users`
> 
> Screenshot 2019-10-09 at 09 50 35
> 
> I think that generally it is a good practice to have an index on fk columns.

Turns out this branch is relevant, and I've pushed the new commit with your suggestion. Thanks!

**Comment by pollygee** [comment]
@abronkema I think we want to ignore Organization at the moment, because we only have 1 and not solve for this problem we may have in the future.  I don't want to couple user and organization at all.

I think we could link a user to a location (places where they meet which each has a different set of inventory), but we should check with MAB because if they want a user to have access to all the inventory in all the locations, then we probably don't need to link a user to a location.  Let's chat more about this issue.

**Comment by LizPrescott** [comment]
Closing this to keep organizations out of the picture for now.
